### ISSUE TITLE
Elliptic Curve algebra hierarchy proposal

### ISSUE THREAD

**Issue by echatav** [zkFold/zkfold-base#411]
Subsequent to #404 assuming that the type family `BooleanOf` has been moved from the `EllipticCurve` class to the `Conditional` class, it will make sense to also reorganize the elliptic curve classes & datatypes which I have worked on in #396. The two "classes" of elliptic curves all of our examples (so far) fall into are either standard Weierstrass curves or [twisted Edwards curves](https://eprint.iacr.org/2008/013.pdf). Here is the proposal sketch:

```Haskell
class Planar point where
  pointXY :: field -> field -> point field

class ProjectivePlanar point where
  pointInf :: point

class
  ( Field field
  , Conditional field
  , Planar (point curve)
  , AdditiveGroup (point curve field)
  ) => EllipticCurve point curve field where
    isOnCurve :: point curve field -> BooleanOf field

class
  ( EllipticCurve point curve baseField
  , FiniteField baseField
  , FiniteField scalarField
  , Scale scalarField (point curve baseField)
  ) => SubgroupCurve point curve baseField scalarField where
    pointGen :: point curve baseField

class EllipticCurve point curve field
  => SymmetricCurve pt point curve field | point -> pt, pt -> point where
    pointCompressed :: field -> BooleanOf field -> pt curve field
    compressPoint :: point curve field -> pt curve field
    decompressPoint :: pt curve field -> point curve field

class EllipticCurve point curve field
  => WeierstrassCurve point curve field where
    weierstrassA :: field
    weierstrassB :: field

class EllipticCurve point curve field
  => TwistedEdwardsCurve point curve field where
    twistedEdwardsA :: field
    twistedEdwardsD :: field

data Point curve field = Point
  { _x :: field
  , _y :: field
  , _isInf :: BooleanOf field
  } deriving Generic

data CompressedPoint curve field = CompressedPoint
  { _x :: field
  , _bigY :: BooleanOf field
  , _isInf :: BooleanOf field
  } deriving Generic

data AffinePoint curve field = AffinePoint
  { _x :: field
  , _y :: field
  } deriving Generic

data CompressedAffinePoint curve field = CompressedAffinePoint
  { _bigX :: BooleanOf field
  , _y :: field
  } deriving Generic

newtype Weierstrass point curve field =
  Weierstrass {pointWeierstrass :: point curve field}

newtype TwistedEdwards point curve field =
  TwistedEdwards {pointTwistedEdwards :: point curve field}
```

Classes maximize polymorphism over different field and point types. All point types are generic records of kind `(curve :: k) -> (field :: Type) -> Type` so they work in Haskell or Symbolic. When defining instances for a curve, you just fill in the values for the `pointGen` method of `SubgroupCurve` and the method parameters of either `WeierstrassCurve` or `TwistedEdwardsCurve` and then you should be able to derive `AdditiveMonoid`, `Scale`, `EllipticCurve`, `SymmetricCurve` etc. instances either via `Weierstrass` or `TwistedEdwards` newtypes. Point types can be symbolic inputs with `isValid = isOnCurve`.

notes:
* the reason the compressed affine point is y-axis symmetric is because of the twisted Edwards form of the curve. Weierstrass curves are x-axis symmetric.
* We could also intro special case where `weierstrassA = 0` which is true for all of our examples and simplifies point doubling and `isOnCurve` slightly.
* We can also introduce new point types like homogeneous coordinates and twisted coordinates.
* The `curve` type parameter is completely phantom. And unlike the current situation a Symbolic `curve` doesn't need to know its `Context` since the choice of `field` will. Thus `curve` can have any kind. I recommend having `curve :: Symbol` and using strings like `"ed25519"` and `"secp256k1"` etc. Then you don't need to define a datatype for the curve, it's just its name.

**Comment by vlasin** [comment]
A note on elliptic curves in Symbolic.

Generally speaking, we can define a circuit over any field. However, in practice, we are constrained by what we want to do with it. Most proving protocols require that the circuit's field is a scalar field of some cyclic group.

There are three possibilities for how the circuit's field can relate to the EC fields:

1. Both the base and scalar fields of the curve are foreign.
2. The circuit's field coincides with the scalar field.
3. The circuit's field coincides with the base field.

Options 1 and 2 make sense for any curve. Option 3 only makes sense for a pair of curves that form a cycle (e.g., Pasta curves) when the base field of one curve is the scalar field of another. So, for each curve we add to Symbolic, we should have 2 or 3 options available.

**Comment by echatav** [comment]
> So, for each curve we add to Symbolic, we should have those two or three options available.

In the current curve class, the fields are associated type families. That means you only get 1 option for the field. By making the curve classes be multiparameter and parametric over the choice of field, we can instantiate different field type options that make sense across both Haskell & Symbolic.

**Comment by echatav** [comment]
It's possible to simplify the proposal if we _assume_ that `Point` is always on a Weierstrass curve and `AffinePoint` is always on a twisted Edwards curve. Then we don't need the `Weierstrass` & `TwistedEdwards` newtypes. We just need these instances instead:

```
instance WeierstrassCurve Point curve field => EllipticCurve Point curve field
instance WeierstrassCurve Point curve field => SymmetricCurve CompressedPoint Point curve field
instance WeierstrassCurve Point curve field => AdditiveSemigroup (Point curve field)
instance WeierstrassCurve Point curve field => AdditiveMonoid (Point curve field)
instance WeierstrassCurve Point curve field => AdditiveGroup (Point curve field)
instance WeierstrassCurve Point curve field => Scale Natural (Point curve field)
instance WeierstrassCurve Point curve field => Scale Integer (Point curve field)
```
And the similar for `EdwardsCurve AffinePoint`.

**Comment by vlasin** [comment]
Seems reasonable. Adding @TurtlePU to the conversation.

**Comment by TurtlePU** [comment]
Hierarchy in general looks great! A couple questions

Is it possible to define `point` in `Planar` to be just a `Type`, not a type operator? Like
```haskell
class Planar field point | point -> field where
  fieldXY :: field -> field -> point
```
And same for `*Curve` classes. The reason I'm asking is the case when, for example, we might want to introduce some highly efficient point type which is specific for some curve and field, rather than a general datatype

**Comment by echatav** [comment]
^ I think both `Planar` and `ProjectivePlanar` can be subject to change. When you introduce a highly efficient type we can alter them accordingly. For now their designs are such that `pointXY`, `pointInf`, and `pointCompressed` constructors work (or don't) sensibly across the point types `Point`, `CompressedPoint`, `AffinePoint` & `CompressedAffinePoint`.

**Comment by echatav** [comment]
One example of an optimized point type would be [extended twisted Edwards coordinates](https://eprint.iacr.org/2008/522.pdf) but these too are just a generic 4-tuple of fields.

**Comment by echatav** [comment]
`Symmetric pt point curve field` should imply `Ord field` since we're doing comparisons. Moreover, it's possible to define `CompressedPoint` like so!

```Haskell
data CompressedPoint curve field = CompressedPoint
  { _x :: field
  , _ltlY_inf_bigY :: OrderingOf field
  } deriving Generic
```

**Comment by echatav** [comment]
Point compression is a little more complicated than I thought. The parity bit is calculated differently for different curves like `secp256k1`, `bls12-381` and `ed25519` and so compression/decompression is not always the same algorithm even among Weierstrass curves.

**Comment by echatav** [comment]
Naming question: Should `ProjectivePlanar` just be `Projective`? Any dimension projective space has a point at infinity, `pointInf`. The only reason I don't like it is that "projective" is so overloaded terminology but I guess within the curve hierarchy & module the context is clear. Maybe it should be `PointAtInfinity`?

**Comment by vlasin** [comment]
I guess `PointAtInfinity` is fine.

**Comment by echatav** [comment]
```Haskell
data ExtendedPoint curve field = ExtendedPoint
  { _x :: field
  , _y :: field
  , _z :: field
  } deriving Generic
instance Field field => PointAtInfinity (ExtendedPoint curve field) where
  pointInf = ExtendedPoint 0 1 0
  
data TwistedExtendedPoint curve field = TwistedExtendedPoint
  { _x :: field
  , _y :: field
  , _z :: field
  , _t :: field
  } deriving Generic
```
^ both of these are potentially useful/optimized point (coordinates) types that can be instantiated for twisted Edwards curves besides `AffinePoint` (and `CompressedAffinePoint`).

**Comment by echatav** [comment]
Instead of incurring being blocked by reorganizing logical hierarchy as in #404 and #421 one could add a `bool` parameter to this proposal. This construction feels a bit more complex but also more consistent. Should I do this instead?

```Haskell
class Planar point where
  pointXY :: field -> field -> point field

class ProjectivePlanar point where
  pointInf :: point

class
  ( Field field
  , Conditional bool field
  , Planar (point curve)
  , AdditiveGroup (point curve field)
  ) => EllipticCurve point bool curve field where
    isOnCurve :: point curve field -> bool

class
  ( EllipticCurve point bool curve baseField
  , FiniteField baseField
  , FiniteField scalarField
  , Scale scalarField (point curve baseField)
  ) => SubgroupCurve point bool curve baseField scalarField where
    pointGen :: point curve baseField

class EllipticCurve (point bool) bool curve field
  => SymmetricCurve pt point bool curve field | point -> pt, pt -> point where
    pointCompressed :: field -> bool -> pt bool curve field
    compressPoint :: point bool curve field -> pt bool curve field
    decompressPoint :: pt bool curve field -> point bool curve field

class WeierstrassCurve curve field where
  weierstrassA :: field
  weierstrassB :: field

class TwistedEdwardsCurve curve field where
  twistedEdwardsA :: field
  twistedEdwardsD :: field

data Point bool curve field = Point
  { _x :: field
  , _y :: field
  , _zBit :: bool
  } deriving Generic

data CompressedPoint bool curve field = CompressedPoint
  { _x :: field
  , _yBit :: bool
  , _zBit :: bool
  } deriving Generic

data AffinePoint curve field = AffinePoint
  { _x :: field
  , _y :: field
  } deriving Generic

data CompressedAffinePoint bool curve field = CompressedAffinePoint
  { _bitX :: bool
  , _y :: field
  } deriving Generic

data HomogeneousPoint curve field = HomogeneousPoint
  { _x :: field
  , _y :: field
  , _z :: field
  } deriving Generic

data TwistedExtendedPoint curve field = TwistedExtendedPoint
  { _x :: field
  , _y :: field
  , _z :: field
  } deriving Generic


newtype Weierstrass point curve field =
  Weierstrass {pointWeierstrass :: point curve field}

newtype Montgomery point curve field =
  Montgomery {pointMontgomery :: point curve field}

newtype TwistedEdwards point curve field =
  TwistedEdwards {pointTwistedEdwards :: point curve field}
```

**Comment by echatav** [comment]
I will start on this soon. With this modified proposal, elliptic curve hierarchy is not blocked by logical hierarchy changes (I think) so we can continue discussing the latter.

**Comment by TurtlePU** [comment]
> ```haskell
>     pointCompressed :: field -> bool -> pt bool curve field
>     compressPoint :: point bool curve field -> pt bool curve field
> ```

I think that instead of being included as part of the point type, `bool` has to be inferred from `field` (this would happen anyway because `EllipticCurve` instances would surely need `(==) :: field -> field -> bool`)

Also I'd like to mention that this decision of whether to include `bool` in the point type or not would be deferred to the instance implementation stage if classes were defined for `point :: Type` instead of `point :: Type -> k -> Type -> Type` or something like this.

**Comment by vlasin** [comment]
> I think that instead of being included as part of the point type, bool has to be inferred from field (this would happen anyway because EllipticCurve instances would surely need `(==) :: field -> field -> bool)`

Meaning that if we are working with a symbolic point type, we get the corresponding symbolic bool, and if we are working with, say, `Zp p`, we get the Haskell bool?

**Comment by TurtlePU** [comment]
> Meaning that if we are working with a symbolic point type, we get the corresponding symbolic bool, and if we are working with, say, `Zp p`, we get the Haskell bool?

Exactly!
